### PR TITLE
Dynamic pines url

### PR DESCRIPTION
**Pull request by Kayan1920**
- PINES compatibility with superbio server using token authentication.
- Updates to the UI.
- Fixed bug regarding index updates.

### CODE REVIEW

File: cedars/app/api.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    - Call this URL for PINES predictions
+    '''
+
+    env_url = os.getenv("PINES_API_URL")
+    api_url = os.getenv("SUPERBIO_API_URL")
+    if env_url is not None:
+        # Get PINES api from .env
+        pines_api_url = env_url
+        is_url_from_api = False
+        logger.info(f"Received url : {pines_api_url} for pines from ENV variables.")
```
Comment (ygivenx) [inline cedars/app/api.py:22]: in this case, can you hit the `/healthcheck` endpoint in PINES to make sure it is up and running? 

@app.get("/healthcheck")
async def healthcheck():
    return {"message": f"Welcome to the Pines NLP Model\n\n Current Model: {model_name}",
            "status": "Healthy"}

File: cedars/app/api.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
+import os
+from loguru import logger
+from tenacity import retry, wait_exponential
+import requests
+from . import db
+
+def load_pines_url(project_id, superbio_api_token = None):
```
Comment (ygivenx) [inline cedars/app/api.py:7]: can you define the expected return type in the function signature and also add documentation.

File: cedars/app/api.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        try:
+            logger.info("\n\nPinging", f'{api_url}/{endpoint}')
+            logger.info("With header : ", headers, flush=True)
+            response = requests.post(f'{api_url}/{endpoint}', headers=headers, data={})
+            logger.info("POST responce", response, flush=True)
+
+            if response.status_code != 200:
+                raise requests.exceptions.HTTPError
+        except Exception as e:
+            logger.error(f"Encountered error {e} when trying to start PINES server")
```
Comment (ygivenx) [inline cedars/app/api.py:43]: you should just `raise` the exception and put the calling function in try catch?

File: cedars/app/api.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    logger.info("\n\nGot JSON", json_data, flush=True)
+    return json_data['url']
+
+def kill_pines_api(project_id, superbio_api_token):
+    '''
+    Shutsdown remote PINES server if it is running.
+    Currently only applicable when using the superbio API system.
+    '''
+
+    if db.is_pines_api_running() and superbio_api_token is not None:
```
Comment (ygivenx) [inline cedars/app/api.py:78]: what happens if superbio token expires?

File: cedars/app/auth.py [hunk 5]
```diff
@@ -68,26 +77,40 @@ def __init__(self, data):
...
             error = 'Passwords do not match.'
         elif not username or not password or len(username.strip()) == 0 or len(password.strip()) == 0:
             error = "Username and password are required."
-
         elif existing_user or username.lower() in ['cedars', 'pines']:
             error = 'Username already exists or reserved. Please choose a different one.'
+        elif password_result is False:
+            # If the password is not in compliance with our policy,
+            # display the first issue found with this password
+            error = password_issues[0]
```
Comment (ygivenx) [inline cedars/app/auth.py:113]: you can display all issues at once by doing something like `"\n".join(password_issues)`

File: cedars/app/auth.py [hunk 6]
```diff
@@ -106,12 +129,29 @@ def register():
             flash('Registration successful.')
             if no_admin:
                 login_user(User(db.get_user(username)))
-
+            init_pines()
```
Comment (ygivenx) [inline cedars/app/auth.py:132]: does this mean that the user will stuck during login unless PINES comes up?

File: cedars/app/auth.py [hunk 7]
```diff
@@ -106,12 +129,29 @@ def register():
...
+    api_url = f'{os.getenv("SUPERBIO_API_URL")}/users/{user_id}/cedars_projects/{project_id}'
+    headers = {"Authorization": f"Bearer {token}"}
+    logger.info(f"Verifying token with {api_url}")
+    logger.info(f"Headers: {headers}")
+    try:
+        response = requests.get(api_url, headers=headers, verify=True)
+        if response.status_code == 200:
+            # Assuming the API returns user data on successful verification
+            return response.json()
+        return None
```
Comment (ygivenx) [inline cedars/app/auth.py:150]: you can replace this with

```python
try:
  ....
finally:
  return None
```

File: cedars/app/auth.py [hunk 8]
```diff
@@ -123,6 +163,7 @@ def login():
 
         if user and check_password_hash(user['password'], password):
             login_user(User(user))
+            init_pines()
```
Comment (ygivenx) [inline cedars/app/auth.py:166]: will the user be waiting until pines comes up?

File: cedars/app/auth.py [hunk 9]
```diff
@@ -133,6 +174,49 @@ def login():
...
+        if not existing_user:
+            # Create a new user with data from the external API
+            db.add_user(
+                username=username,
+                password=generate_password_hash(token),  # Store hashed token as password
+                is_admin=True if "admin" in user_data["user"].get('institution_roles') else False
+            )
+        user = User(db.get_user(username))
+        session["superbio_api_token"] = token
+        init_pines(session["superbio_api_token"])
```
Comment (ygivenx) [inline cedars/app/auth.py:208]: will the user be stuck until pines comes up?

File: cedars/app/ops.py [hunk 10]
```diff
@@ -337,6 +344,8 @@ def do_nlp_processing():
     """
     nlp_processor = nlpprocessor.NlpProcessor()
     pt_ids = db.get_patient_ids()
+    superbio_api_token = session.get('superbio_api_token')
```
Comment (ygivenx) [inline cedars/app/ops.py:347]: did you add the expiration logic too?

File: cedars/app/ops.py [hunk 11]
```diff
@@ -345,17 +354,64 @@ def do_nlp_processing():
...
+    A callback function to handle the event where
+    a job from the task queue is completed successfully.
+    '''
+    db.report_success(job)
+    
+    queue_length = len(flask.current_app.task_queue)
+    all_jobs = StartedJobRegistry(queue=flask.current_app.task_queue)
+    num_running_jobs = len(all_jobs.get_job_ids())
+
+    if queue_length == 0 and num_running_jobs == 1:
```
Comment (ygivenx) [inline cedars/app/ops.py:380]: we need to discuss this

File: cedars/app/ops.py [hunk 12]
```diff
@@ -569,16 +633,17 @@ def adjudicate_records():
...
-                if patient:
-                    patient_id = patient["patient_id"]
+            search_patient = convert_to_int(search_patient)
+            patient = db.get_patient_by_id(search_patient)
+            if patient is None:
+                patient_id = None
+            else:
+                is_patient_locked = db.get_patient_lock_status(search_patient)
+                if is_patient_locked is False:
+                    if patient:
```
Comment (ygivenx) [inline cedars/app/ops.py:643]: if we are in `else` of `patient is None` do we need to this `if`?

File: nginx/nginx.conf [hunk 13]
```diff
@@ -55,18 +55,19 @@ http {
...
-        proxy_set_header Host $http_host;
-        proxy_set_header X-Real-IP $remote_addr;
-        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
-        proxy_set_header X-Forwarded-Proto $scheme;
-
-        location / {
-            proxy_pass http://pines:8036;
-        }
-    }
+    # server {
```
Comment (ygivenx) [inline nginx/nginx.conf:58]: please uncomment this - we can't push this to main

**Review by ygivenx — CHANGES_REQUESTED** [review]
Please make the changes - do not push your changes before meeting

**Review by ygivenx — CHANGES_REQUESTED** [review]
Please address all comments and answer or resolve them

File: cedars/app/auth.py (hunk 7 above)
Comment (Kayan1920) [reply cedars/app/auth.py]: The finally keyword does not work in this use case as the function always returns None.

File: cedars/app/api.py [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+    
+    Returns :
+        bool : True if the token has expired, 
+                False if the token is still valid.
+        
+        None : Returns None if the input is not a valid token.
+
+    '''
+    api_url = os.getenv("SUPERBIO_API_URL")
+    if api_url is None:
```
Comment (ygivenx) [inline cedars/app/api.py:123]: condense it to one line

Comment (ygivenx) [reply cedars/app/api.py:123]: return type - organize

File: cedars/app/auth.py [hunk 15]
```diff
@@ -133,6 +170,48 @@ def login():
...
+                is_admin=True if "admin" in user_data["user"].get('institution_roles') else False
+            )
+        user = User(db.get_user(username))
+        session["superbio_api_token"] = token
+        if 'project' in user_data and 'model' in user_data['project']:
+            # If we are are using any model with this login,
+            # then we start the PINES server and get it's URL
+            pines_url, is_url_from_api = load_pines_url(project_id,
+                                        superbio_api_token=token)
+            db.create_pines_info(pines_url, is_url_from_api)
```
Comment (ygivenx) [inline cedars/app/auth.py:209]: please make sure that the user is not waiting for the tenacity to complete retrying

File: cedars/app/db.py [hunk 16]
```diff
@@ -175,6 +186,15 @@ def create_index(collection, index: list):
             mongo.db[collection].create_index(i)
         logger.info(f"Created index {i} in collection {collection}.")
 
+def create_db_indices():
+    '''
+    Creates indices for the ANNOTATIONS and PINES cols in the db.
+    '''
+    logger.info("All tasks completed.")
+    logger.info("Creating indexes for ANNOTATIONS.")
+    create_index("ANNOTATIONS", ["patient_id", "note_id"])
```
Comment (ygivenx) [inline cedars/app/db.py:195]: please test with real upload

File: cedars/app/db.py [hunk 17]
```diff
@@ -1379,22 +1431,41 @@ def update_db_task_progress(task_id, progress):
...
+    the PINES server if it is running via a SUPERBIO api.
     """
     job.meta['progress'] = 0
     job.save_meta()
+    # If the TASK queue does not have any jobs :
+    # 1. Start indexing the annotations and notes db
+    queue_length = len(flask.current_app.task_queue)
+    if queue_length == 0:
+        # Index database if all tasks are completed
+        create_db_indices()
```
Comment (ygivenx) [inline cedars/app/db.py:1468]: please test with real upload

File: cedars/app/ops.py [hunk 18]
```diff
@@ -345,17 +387,100 @@ def do_nlp_processing():
...
+        - superbio_api_token (str) : Access token for superbio server if one is being used.
+    
+    Returns :
+        (bool) : True if a valid pines url has been found.
+                            False if not valid pines url available.
+    '''
+    project_info = db.get_info()
+    project_id = project_info["project_id"]
+
+    if superbio_api_token is None:
```
Comment (ygivenx) [inline cedars/app/ops.py:455]: redundant code

**Review by ygivenx — CHANGES_REQUESTED** [review]

File: cedars/app/api.py [hunk 19]
```diff
@@ -0,0 +1,182 @@
...
+    Args :
+        - project_id (str) : The ID of the current CEDARS project.
+        - superbio_api_token(str / None) : API token for the superbio server running PINES.
+    '''
+
+    if db.is_pines_api_running() and superbio_api_token is not None:
+        # kill PINES server if using superbio API
+        logger.info("Killing PINES server.")
+        api_url = os.getenv("SUPERBIO_API_URL")
+        if api_url is not None:
```
Comment (ygivenx) [inline cedars/app/api.py:170]: this condition is checked already

File: cedars/app/api.py [hunk 20]
```diff
@@ -0,0 +1,182 @@
...
+            endpoint = f"api/cedars_projects/{project_id}/pines"
+            if superbio_api_token is not None:
+                headers = {"Authorization": f"Bearer {superbio_api_token}"}
+            else:
+                headers = {}
+            try:
+                requests.delete(f'{api_url}/{endpoint}', headers=headers)
+            except Exception as e:
```
Comment (ygivenx) [inline cedars/app/api.py:178]: add TODO: handle different exceptions.

**Review by ygivenx — CHANGES_REQUESTED** [review]

**Review by ygivenx — APPROVED** [review]

**Pull request merged**